# Question 1:
Use while only when the number of iterations is indefinite.

If you can term the steps as do n times, do once for each item, etc. use for loop instead.

If you can only term the steps as do until something happens. Like when user inputs 10.

A bit of wisdom 📖 There are maily two ways in which while loops are used in the context of taking inputs until a terminal word.

Method 1
Method 2
# method 1
a = input()

while a != terminal_word: # opposite of the terminal condition
    # do something with a
    a = input() # take the next a

# method 2
while True: # loop forever
    a = input()
    if a == terminal_word: # the terminal condition
        break
    # do something with a
Problem Statement

Problem type - Standard Input - Standard Output

NOTE: None of this problem statements can be written using a for since the number of repetition is indefinite.

Implement different parts of a multi-functional program based on an initial input value. Each part of the program will handle various tasks related to accumulation, filtering, mapping, and combinations of these operations. None of the tasks should use explicit loops for definite repetitions, and the program should handle indefinite inputs gracefully.

Tasks

Accumulation - Accumulating a final result

sum_until_0: Continuously read integers from standard input until you receive a zero. Print the sum of these integers.

total_price: Continuously read pairs of integers from standard input, representing the quantity and price of items, until you receive the string "END". Print the total price of all items.

Filtering - Selecting based on a criterion

only_ed_or_ing: Continuously read strings from standard input until you encounter the word "STOP" (case insensitive and not included in the output). Print only those strings that end with "ed" or "ing" (case insensitive).

reverse_sum_palindrome: Continuously read positive integers from standard input until you encounter a "-1"(not included in the output). Print only those integers for which the sum of the number and its reverse is a palindrome.

Mapping - Applying the same operation to different items

double_string: Continuously read lines from standard input until an empty line is encountered. Print each line repeated twice.

odd_char: Continuously read strings from standard input until you encounter a string ending with a "."(include that string with the "." in the output). Extract characters at odd positions (starting from 1) of each line, and print the results in a single line separated by spaces.

Filter and Map - Applying an operation to selected items

only_even_squares: Continuously read numbers from standard input until "NAN" is encountered. Print the square of each number only if it is even.
Filter and Accumulate - Accumulating a result with selected items

only_odd_lines: Continuously read lines from standard input until "END"(not included in the output) is encountered. Create a string by prepending only the odd lines (starting from 1) with a newline character in between, and print the result which will be the odd lines in reverse order.

In [ ]:
# Note this prefix code is to verify that you are not using any for loops in this exercise. This won't affect any other functionality of the program.
#with open(__file__) as f:
#    content = f.read().split("# <eoi>")[2]
#if "for " in content:
#    print("You should not use for loop or the word for anywhere in this exercise")

# This is the first line of the exercise
task = input("Enter your task: ")
# <eoi>

if task == "sum_until_0":
  total = 0
  num = int(input("Enter an integer here: "))
  while (num != 0):
    total = total + num
    num = int(input("Enter an integer here: "))
  print(total)

elif task == "total_price":
  total_price = 0
  while True:
    line = input("Enter integers here with a space in it: ")
    if (line == "END"):
      break
    quantity, price = line.split() # it splits the input into quantity and price.
    quantity, price = int(quantity), int(price)
    total_price = total_price + quantity * price
  print(total_price)

elif task == "only_ed_or_ing":
  word = input("Enter a word here: ")
  while True:
    if (word == "STOP"):
      break
    if (word.lower().endswith('ed') or word.lower().endswith('ing')):
      print(word)
    else:
      pass
    word = input("Enter a word here: ")

elif task == 'reverse_sum_palindrome':
  num = int(input("Enter an integer here: "))
  while (num != -1):
    rev_num = int(str(num)[::-1])
    num_sum = num + rev_num
    reverse = int(str(num_sum)[::-1])
    if (num_sum == reverse):
      print(num)
    else:
      pass
    num = int(input("Enter an integer here: "))

elif task == 'double_string':
  word = input("Enter a word here: ")
  while (word != ''):
    word = word + word
    print(word)
    word = input("Enter a word here: ")

elif task == 'odd_char':
  line = input("Enter a word here: ")
  while (line[:-1] != '.'):
    odd_char = line[::2]
    print(odd_char, end = '')
    line = input("Enter a word here: ")
  print(line[::2])

elif task == 'only_even_squares':
  while True:
    line = input('Enter a value here: ')
    if (line == "NAN"):
      break
    num = int(line)
    square_value = num ** 2
    if (num % 2 == 0):
      print(square_value)

elif task == 'only_odd_lines':
  line = input("Enter a word here: ")
  new_line = ''
  while (line != "END"):
    new_line += line + ' '
    line = input("Enter a word here: ")
  split_line = new_line.split()
  odd_lines = split_line[::2]
  reversed_odd_line = odd_lines[::-1]
  reverse_line = '\n'.join(reversed_odd_line)
  print(reverse_line)

Enter your task: only_odd_lines
Enter a word here: one
Enter a word here: two
Enter a word here: three
Enter a word here: END
three
one


In [ ]:
n = int(input("Enter an integer here: "))
total = 0
while (n != 0):
  total = total + n
  n = int(input("Enter an integer here: "))
print(total)

Enter an integer here: 10
Enter an integer here: -5
Enter an integer here: 5
Enter an integer here: 0
10


In [ ]:
total_price = 0
while True:
  line = input("Enter the values: ")
  if (line == "END"):
    break
  quantity, price = line.split()
  quantity, price = int(quantity), int(price)
  total_price = total_price + (quantity * price)
print(total_price)

Enter the values: 5 10
Enter the values: 2 20
Enter the values: END
90


In [ ]:
"""
word = input("Enter a word here:")
while (word != "STOP"):
  if (word.lower().endswith('ed') or word.lower()[-3:] == 'ing'): # we can do any of the single way possible...
    print(word)
  word = input("Enter a word here:")
                                                                                                                            """
  #                                                                     OR

word = input("Enter a word here: ")
while (word != 'STOP'):
  if (word.endswith('ed') or word.endswith('ED') or word.endswith('ing') or word.endswith('ING')):
    print(word)
  else:
    pass
  word = input("Enter a word here: ")

Enter a word here: Twinkling
Twinkling
Enter a word here: printING
printING
Enter a word here: Farmed
Farmed
Enter a word here: HARMED
HARMED
Enter a word here: STOP


In [ ]:
"""
num = int(input("Enter an integer here: "))
while (num != -1):
  rev_num = int(str(num)[::-1])
  num_sum = num + rev_num
  if (str(num_sum) == str(num_sum)[::-1]):
    print(num)
  num = int(input("Enter an integer here: "))
                                                                                    """
#                                                           OR

n = int(input("Enter an integer here: "))
while (n != -1):
  if (n > 0):
    rev = int(str(n)[::-1])
    sum = n + rev
    rev_sum = int(str(sum)[::-1])
    if (sum == rev_sum):
      print(n)
    else:
      pass                          # this pass is for exiting this following if block....
  else:
    pass                  # This pass is for exiting this first if block...
  n = int(input("Enter an integer here: "))

Enter an integer here: 56
56
Enter an integer here: 32
32
Enter an integer here: 99
Enter an integer here: -1


In [ ]:
"""
word = input("Enter a word here: ")
while (word != ''):
  print(word * 2)
  word = input("Enter a word here: ")
                                                    """
#                                      OR
word = input("Enter a word here: ")
while (word != ''):
  repeat_word = word + word
  print(repeat_word)
  word = input("Enter a word here: ")


Enter a word here: tan
tantan
Enter a word here: pan
panpan
Enter a word here: 


In [ ]:
"""
line = input()
while line[-1] != ".":
  print(line[::2],end=" ")
  line = input()
print(line[::2])
                                                                            """
#                                                 OR

line = input()
while(line != line.endswith('.')):
  odd_car = line[::2]
  print(odd_car, end = " ")
  line = input()


In [ ]:
"""
num = input()
while(num != "NAN"):
  num = int(num)
  square = (num * num)
  if (num % 2 == 0):
    print(square)
  else:
    pass
  num = input()
                                                                      """
#                                                                             OR


while True:
  line = input()
  if (line == "NAN"):
    break
  num = int(line)
  if (num % 2 == 0):
    square = num ** 2
    print(square)


3
4
16
5
NAN


In [ ]:
"""
new_line = ''
line = input()
while (line != 'END'):
  new_line += line + ' '
  line = input()
split_line = new_line.split()
odd_line = split_line[::2]
reverse_odd_line = odd_line[::-1]
reversed_line = '\n'.join(reverse_odd_line)
print(reversed_line)
                                                    """
#                                                                     OR

result = ''
line_no = 1
while True:
  line = input()
  if (line == "END"):
    break
  if (line_no % 2 != 0):
    result = line + '\n' + result
  line_no += 1
print(result)

one
two 
three
four
five
END
five
three
one



# Question 2:
Write a multi functional program that takes input task from standard input and does the corresponding taks accordingly. Note that the useage of for loop is not allowed in this exercise.

Part 1 - while loop to for loop
factorial - print factorial of a given non-negative integer n (Type: Accumulation)

Input - n:int

even_numbers - Print the even numbers from 0 (including) till the given input number n(including) in multiple lines (Type: Just Iterating)

Input - n:int

power_sequence - Print the sequence 1, 2, 4, 8, 16, ... n terms in same line in multiple lines, where n is taken from the input(Type: Mapping)

Input - n:int

Part 2 - for loop With range

sum_not_divisible - Print the sum of positive less that the given number n and not divisible by 4 and 5. (Type: Filtered Accumulation)

Input - n:int

from_k - Starting from 100 and going in the decreasing order, print the reverse(digits reversed) of first n numbers starting from k which do not have the digit 5 and 9 and is odd number in multiple lines.

Input - n:int, k:int

part 3 - for loop with iterables.

string_iter - Given a string s of digits print the numerical value of the digit multiplied by the previous digit. Assume the pevious digit for the first element to be 1.

Input - s:str

list_iter - Print the elements of a list l line by line in the format {element} - type: {type} where the element is the current element being iterated by the for loop and type is the type of the element. (Even though list are not covered in this week, this is included to demonstrate the similarity between iterating characters in a str and items in a list)

Input - l:list

In [ ]:
# find the factorial of a given input using for loop only...
num = int(input("Enter an integer here: "))
fact = 1
for i in range(1,num+1):
  fact = fact * i
print(fact)

Enter an integer here: 7
5040


In [ ]:
# Find even numbers from 0 till n in multiple lines using for loop.
num = int(input("Enter an integer here: "))
for i in range(0,num+1,2):
  print(i)

Enter an integer here: 12
0
2
4
6
8
10
12


In [ ]:
# Print the power sequence in mulitple line .. eg. 1,2,4,8,16,...n
num = int(input("Enter an integer here: "))
Power_seq = 1
for i in range(num):
  print(Power_seq)
  Power_seq *= 2

Enter an integer here: 5
1
2
4
8
16


In [ ]:
# Find the +ve integers which are not divisible by 4 and 5 less than n using for loop..
n = int(input("Enter an integer here: "))
total = 0
for i in range(1,n):
  if (i % 4 != 0 and i % 5 != 0):
    total += i
print(total)


Enter an integer here: 2
1


In [ ]:
n = int(input("Enter an integer here: "))
k = int(input("Enter an integer here: "))
count = 0
for i in range(k,0,-1):
  i_str = str(i)
  if ('5' not in i_str and '9' not in i_str and i % 2 != 0):
    print(i_str[::-1])
    count += 1
    if (count == n):
      break

Enter an integer here: 5
Enter an integer here: 99
78
38
18
77
37


In [ ]:
s = input("Enter a value here: ")
prev = 1
for char in s:
  num = int(char)
  print(num * prev)
  prev = num

Enter a value here: 1234
1
2
6
12


# Question 3:

Create a multi-functional program that performs different tasks based on the user input. The program should support the following tasks:

Permutation (permutation): Given a string s, print all the possible two-letter permutations(without repitition) of the letters in the string.

Sorted Permutation (sorted_permutation): Given a string s, print all the possible two-letter permutations(without repetition) of the letters in the string where the first character comes before the second one in alphabetical order. The order in which the permutations are printed is same as the previous one (Type: Filtering).

Repeat the Repeat (repeat_the_repeat): Given a number n, print the numbers from 1 to n in the same line and repeat this n times.

Repeat Incrementally (repeat_incrementally): Given a number n, print a pattern where the k-th line contains the first k numbers and there are n lines in total. For example, if n is 4, the output should be:

1

12

123

1234

Increment and Decrement (increment_and_decrement): Given a number n, print a pattern where the k-th line should have the numbers from 1 to k and then back down to 1. For example, if n is 4, the output should be:

1

121

12321

1234321

In [9]:
task = input("Enter a task here: ")

if task == "permutation":
  s = input("Enter a string here: ")
  for i in range(len(s)):
    for j in range(len(s)):
      if (i != j):
        print(s[i] + s[j])

elif task == "sorted_permutation":
  s = input("Enter an input here: ")
  for i in range(len(s)):
    for j in range(len(s)):
      if (i != j and s[i] < s[j]):
        print(s[i] + s[j])

elif task == "repeat_the_repeat":
  n = int(input("Enter an integer here:"))
  for _ in range(n):
    for i in range(1,n+1):
      print(i,end = '')
    print()

elif task == "repeat_incrementally":
  n = int(input("Enter an integer here: "))
  for i in range(1,n+1):
    for j in range(1,i+1):
      print(j,end = '')
    print()

elif task == "increment_and_decrement":
  n = int(input("Enter an integer here: "))
  for i in range(1,n+1):
    for j in range(1,i+1):
      print(j,end = '')
    for j in range(i-1,0,-1):
      print(j,end = '')
    print()

else:
  print("Invalid")

Enter a task here: increment_and_decrement
Enter an integer here: 4
1
121
12321
1234321


#**Question 4:**

You are tasked with writing a program that can handle various tasks based on the input. The first line of the input represents the task to be performed. The possible tasks are:

factors - Find the factors of a number n (including 1 and itself) in ascending order.

find_min - Take n numbers from the input and print the minimum number.

prime_check - Check whether a given number is prime or not.

is_sorted - Check if all characters of the given string from input are in alphabetical order. Print the output as "True" or "False" accordingly.

any_true - Take n numbers from input and check if any of the numbers are divisible by 3. Print the output as "True" or "False" accordingly.

manhattan - Take inputs directions such as "UP", "DOWN", "LEFT" and "RIGHT" from the input until the input is "STOP". Assume you are starting from (0,0) in a cartesian coordinate. Find the Manhattan distance between the starting point and the ending point by following the steps in the cartesian plane.
Write a program to solve these tasks. Use loops where necessary.

In [11]:
task = input("Enter a task here: ")

if task == "factors":
  n = int(input('Enter an integer here: '))
  for i in range(1,n+1):
    if (n % i == 0):
      print(i)

elif task == ""

Enter a task here: factors
Enter an integer here: 35
1
5
7
35
